### ISSUE TITLE
API for Creating Custom Gutters

### ISSUE THREAD

**Issue by jssln** [atom/atom#5038]
## Motivation ##
The Decorations API allows you to attach a class to gutters (and other things) and thus style them. However, we are working on a feature that requires us to be able to add our own HTML elements to the gutter, and would like to propose an API to do that.

**Feature:**
We would like to add buttons to the gutter next to lines that contain a test definition. Clicking on the button would run that test. (This is similar to functionality in Xcode.)

**Why CSS doesn't work:**
Adding CSS only allows us to create “buttons” via :before or :after pseudo-elements. These elements can't respond to events targeted to them; instead, any events they receive will register on the existing “line-number” gutter element. The gutter element already has another action to take in response to a click: it may expand or collapse a section of code. If we tried to add events to the gutter element itself (instead of a new child div), our events would conflict with the existing events from Atom.

**How the Line Numbers in the Gutter Work, Currently:**
When the editor is scrolled, line-number divs that have scrolled offscreen are removed, and new line-number divs for newly onscreen lines are added. The changes are primarily in the GutterComponent class, in “appendOrUpdateVisibleLineNumberNodes” (which calls into “buildLineNumberHTML”) and “removeLineNumberNodes”.

**Proposed Change:**
1. The most flexibility that an Atom API could offer is the ability to accept a function that will be run on each line-number div after it's added (and before it's removed, for any teardown work). In practice, this could lead to changes that affect things that might be considered “core” to Atom's look. It seems like this opens too much flexibility.
2. A more pared-down API that Atom could offer is the ability to add a column next to the gutter (either on the left side or right side, though it's understandable to limit this to the left side because the right side is occupied by “foldability”). The API for a Gutter object would accept a function that is given the line number, and returns a div for that line number, which would be inserted into the DOM next to the appropriate line number. (It would make sense for this insertion to be done in Atom Core code, because it already manages the positioning of the line numbers and does this via absolute positioning for performance optimisation.) 

## Desired Gutter Features ##
**Adding Gutters**
* Create gutter
    * You will most likely create one gutter instance per editor instance.
* API to add a gutter to an editor
    * Indicate whether to position it to the left or right of the line numbers
    * Pass a “z-index” of sorts to indicate how close/far from the edge it should be, similar to the “priority” argument to Atom.Workspace.add---Panel.

**Removing Gutters**
* Remove from view (toggle)
* Destroy

**Gutter Properties to Set**
* Set the width (?)

**Modifying Gutter Contents**
* Insert something inside a div next to each line
    * Initialize the Gutter object with a function that operates on a line number and returns a DOM element. As line-number divs are inserted and removed as the user scrolls, this function is run to determine what to place in the custom gutter next to each line number.
    * API could also allow:
        * Changing this function
        * Clear all the items in the gutter (a convenience)
    * For performance, by default, this function could run just once at initialization on the currently-displayed line numbers, and be run on progressively more line numbers as the user scrolls. The results could be cached.
        * The Gutter API should also expose a method that triggers the function to be run on the currently-displayed line numbers, or on a specific line. This could be useful if you want to update the gutter without the editor being scrolled.

**Comment by mark-hahn** [comment]
This looks good.  I just monkey-patched the DOM to add breakpoints to the
gutter.  So there is at least one more use case.

**Comment by jssln** [comment]
Here's a draft of what an API might look like:

###Gutter###
Don't create an instance of this directly. Instead, go through the TextEditor::addGutter() method.

**::destroy()**
Destroy and remove this gutter from the UI.

**::refreshVisibleLines()**
Refreshes the content of the gutter next to the lines currently displayed, by calling the function provided at initialization.

**::refreshLine(**linenumber: Number**)**
Refreshes the content of the gutter next to the given line, by calling the function provided at initialization.

**::hideContents()**
Hide the contents of the gutter.

**::showContents()**
Show the contents of the gutter.

**::hide()**
Hide this gutter.

**::show()**
Show this gutter.


###TextEditor###
**::addLeftGutter(**function, options**)**
Adds a gutter to the left side of the line numbers. Returns a Gutter instance.

* function(line_number: Number, editor: TextEditor): mixed
    * line_number: the number line that the gutter slot to be decorated is next to
    * editor: the TextEditor that the gutter belongs to
    * This function will be called every time a new line number is scrolled into view.
* options: Object
    * priority: Number
        * Determines stacking order between gutters. Lower priority items are forced closer to the edges of the window. (default: 100)

**::getLeftGutters()**
Returns an Array of all the Gutters to the left side of the line numbers.

**::addRightGutter(**function, options**)**
[methods same as above]

**::getRightGutters()**
[methods same as above]

**Comment by benogle** [comment]
Based on a little discussion with @nathansobo, what we're thinking right now is similar to your suggestions:

### Gutter

Use `TextEditor::add*Gutter()`

**::decorateMarker(marker, options)**
Decorate an area in the gutter corresponding to the marker range. These work similarly to the current `gutter` decorations, with the addition of an `item` option.

* `options` Object
  * `class` (optional)
  * `item` (optional) {Object} model with a corresponding view registered, or an {HTMLElement}. It would use `atom.views.getView()` similar to `workspace.addPanel(item: ...)`

We're debating whether there should be an explicit `type` attr in the options object. Right now it makes sense to be typeless and check `class` or `item`, but maybe some other type we havent thought of uses `class` or `item`?

**::destroy()**
Destroy and remove this gutter from the UI.

**::hide()**
Hide this gutter.

**::show()**
Show this gutter.


### TextEditor
**::addLeftGutter(**options**)**
Adds a gutter to the left side of the line numbers. Returns a Gutter instance.

* `options` {Object}
  * `priority` {Number} Determines stacking order between gutters. Lower priority items are forced closer to the edges of the window. (default: 100)

**::getLeftGutters()**
Returns an Array of all the Gutters to the left side of the line numbers.

**::addRightGutter(**options**)**
[methods same as above]

**::getRightGutters()**
[methods same as above]

**Comment by mark-hahn** [comment]
Is there any ETA on this?  My line-number code I am using has blown up in recent versions.   I'd love to just switch to a new non-monkey-patched solution but I can't wait too long.

**Comment by thedaniel** [comment]
> Is there any ETA on this?

There is not. It looks like a considerable chunk of work, so don't expect it tomorrow ;)  @jssln - were you planning on working on the support for this, or waiting on @atom/core to do it? If the latter, we will probably discuss our roadmap / priorities on a call next week and could have a better idea then.

**Comment by jssln** [comment]
I've been in contact with @benogle about this (with @nathansobo in the loop). I am planning to work on this if it's feasible; we need to work out how to do it to minimize conflict.

**Comment by nathansobo** [comment]
@jssln Are you on board with the decoration-oriented approach to getting content in the gutters that @benogle mentioned above? If so, please ping me on any related PRs or contact me with questions. Editor internals are moving but if you develop good specs I can keep them passing before merging my changes.

**Comment by jssln** [comment]
Yeah, I'm on board. What's the best way to contact you?

**Comment by nathansobo** [comment]
You can ping me via `@nathansobo` on this or any other thread or email me at the email address listed on my GitHub profile.

**Comment by jssln** [comment]
While working on this, I realized we'd also like to specify handlers for when a user interacts with the gutter, in particular by clicking. I.e., when a user clicks on a specific line in the gutter, we want to do something given the gutter and the line number. 

Currently, foldable lines are implemented by a click handler within Atom Core. We could add this to the API either by
a) an explicit method like: **Gutter::onClick(callback(gutter, line-number))**, or 
b) documenting an event that will be triggered by a click, which can then be handled by registering for the event in the CommandRegistry.

I'm guessing (b) is preferable based on the patterns in atom core, but wanted to get your thoughts.

**Comment by nathansobo** [comment]
@jssln Any reason why you couldn't just register the click handler on whatever view you added to the custom gutter? Here's what I'm thinking:

```coffee
# this is using the raw DOM API for creating custom elements. obviously a framework would
# make it clearer
GutterWidgetElement =
  document.registerElement 'gutter-widget', prototype: Object.create HTMLElement.prototype,
    attachedCallback: ->
      @addEventListener 'click', -> # ...

gutter = editor.addGutter()
marker = editor.markBufferRange([[2, 0], [2, Infinity]])
decoration = gutter.decorateMarker(marker, item: new GutterWidgetElement)
```

**Comment by jssln** [comment]
Gotcha. That's a good way.

**Comment by dmnd** [comment]
Another use case for this API is https://github.com/alexcorre/git-blame.

**Comment by kans** [comment]
This would really be a useful addition.  I'd like to drop an img into the gutter which can only be resolved at run time.  The existing solutions are either fragile or extremely heavy handed.

**Comment by mark-hahn** [comment]
There is a new api to do exactly what you want.

**Comment by kans** [comment]
@mark-hahn 
Is the API documented or merged?

**Comment by jssln** [comment]
Sorry, @kans and @mark-hahn, I missed your comments. I'd put this aside for a month, but this is now back to my main priority.

**Comment by jssln** [comment]
This pull request was merged today, so the API will be out in the next release.
https://github.com/atom/atom/pull/5217

**Comment by steelbrain** [comment]
while the API has been implemented, it would've been nice to have `addLeftGutter`

**Comment by nathansobo** [comment]
@steelbrain The problem is that it makes the gutter order dependent on the order in which packages are loaded, which isn't reliable.

**Comment by steelbrain** [comment]
@nathansobo Ideally, we could have the ability to provide html elements for gutters that we can decorate, so when package `a` decides to add a decoration to `gutter .icon-right::before` it doesn't override the styles of package `b`. But then again it would introduce an overlaying problem, so to work around that overlaying problem, we would have to expand the gutter in the left and add all of the elements and give them a small width.

That, in fact, is what some elements are doing right now, examples are wallaby-js. It adds a new gutter element on the left side of the line number one and adds it's decorations there.

**Comment by nathansobo** [comment]
You can add custom gutters now to do that... just have to specify a positive or negative priority depending on your desired side.

**Comment by steelbrain** [comment]
Awesome, I didn't know that, and have been digging through the docs for about an hour for it, no luck. Any link will be appreciated.

**Comment by steelbrain** [comment]
Disregard that, just found it. :facepalm:

**Comment by nathansobo** [comment]
No prob. If you want to send a PR with improvements to the docs based on this bad browsing experience, that would be helpful.
